### PR TITLE
[AI] Another attempt to handle focus switches via target ratio

### PR DESCRIPTION
**Pull request by Grummel7**
This should work better than the old code, but on its own still produces weird decisions. Often it would be better to simply chose always the best focus. That however has two problems:
- Large empires will do almost no research, because when the research priority is low, industry is considered "better" for almost all planets
- When having a lot of similar planets, e.g. mostly starting species without any specials that affect industry or research, a small change in the priorities may make the AI switch all planets at once.

@MichaelCourtney Please have a look at this. The stupid decisions at the beginning could probably be avoided by using more selective priorities. Or perhaps we could use the best only up until some point (minimum number of planets to chose from? Other ideas?)

Also, in general, we should work on the priorities. At least the first point could also be solve in a much simpler way by avoiding extremely low research priorities.

### CODE REVIEW

File: default/python/AI/ResourcesAI.py [hunk 1]
```diff
@@ -520,55 +526,149 @@ def early_capital_handling(self):
...
+        ti_gain = to_industry.difference(INDUSTRY, RESEARCH) if to_industry else USELESS_RATING
+        if tr_gain > 0 and ti_gain > 0:
+            debug("both are better with the new focus")
+            self.bake_future_focus(to_research.planet.id, RESEARCH)
+            self.bake_future_focus(to_industry.planet.id, INDUSTRY)
+        else:
+            pp_per_priority, rp_per_priority = self.get_planned_pp_rp_per_prio()
+            debug(f"current: PP/prio={pp_per_priority:.4f}, RP/prio={rp_per_priority:.4f}")
+            # For an AI without a capital numbers may be 0. But if one is bigger than the other,
+            # at least this one cannot be, so it is safe ot use it as a divisor.
```
Comment (Cjkjvfnby) [inline default/python/AI/ResourcesAI.py]:
```suggestion
            # at least this one cannot be, so it is safe to use it as a divisor.
```

**Review by Cjkjvfnby — APPROVED** [review]
Looks good to me

**Comment by MichaelCourtney** [comment]
Will try it out

**Comment by MichaelCourtney** [comment]
[AI_6.log](https://github.com/freeorion/freeorion/files/12388308/AI_6.log)

The george's 2nd planet keeps switching

**Comment by MichaelCourtney** [comment]
> Also, in general, we should work on the priorities. At least the first point could also be solve in a much simpler way by avoiding extremely low research priorities.

Sounds good to me too, I tend to value both research and industry similarly for most of the game. I don't understand why research priority should ever get that low. It'd also be nice to remove the turn numbers in research priority (turn numbers don't account for non-default settings). One idea might be to base it on a % of total research done instead of turns.

**Comment by MichaelCourtney** [comment]
> [AI_6.log](https://github.com/freeorion/freeorion/files/12388308/AI_6.log)
> 
> The george's 2nd planet keeps switching

This is just the log for 1st 7 turns and you can find the relevant bit by searching for Foci Overview Turn, I'm going to take a look at whats happening when I get home from chess. I suspect the commit is otherwise good, maybe needs some small adjustments to prevent looping?

**Comment by MichaelCourtney** [comment]
Question: Is this waiting on me to test it? I've been a mix of busy and preoccupied. I'll try and find sometime to test it during the week, but otherwise, perfect or not its probably better then what's already in.

**Comment by Grummel7** [comment]
> > [AI_6.log](https://github.com/freeorion/freeorion/files/12388308/AI_6.log)
> > The george's 2nd planet keeps switching
> 
> This is just the log for 1st 7 turns and you can find the relevant bit by searching for Foci Overview Turn, I'm going to take a look at whats happening when I get home from chess. I suspect the commit is otherwise good, maybe needs some small adjustments to prevent looping?

Thanks, I found the bug in my calculations: I did not count already baked planets and since your special handling had baked the capital and the second planet could only produce research or industry so the compared ration were both 0.

I'll have to sleep on this, simply comparing the ratios like this does not work too well.

**Pull request merged**